// @logseq/libs/dist/modules/LSPlugin.Experiments.js
import { safeSnakeCase } from '../helpers';
/**
 * WARN: These are some experience features and may be adjusted at any time.
 * These unofficial plugins that use these APIs are temporarily
 * not supported on the Marketplace.
 */
export class LSPluginExperiments {
    ctx;
    constructor(ctx) {
        this.ctx = ctx;
    }
    get React() {
        return this.ensureHostScope().React;
    }
    get ReactDOM() {
        return this.ensureHostScope().ReactDOM;
    }
    get pluginLocal() {
        return this.ensureHostScope().LSPluginCore.ensurePlugin(this.ctx.baseInfo.id);
    }
    invokeExperMethod(type, ...args) {
        const host = this.ensureHostScope();
        type = safeSnakeCase(type)?.toLowerCase();
        return host.logseq.api['exper_' + type]?.apply(host, args);
    }
    async loadScripts(...scripts) {
        scripts = scripts.map((it) => {
            if (!it?.startsWith('http')) {
                return this.ctx.resolveResourceFullUrl(it);
            }
            return it;
        });
        scripts.unshift(this.ctx.baseInfo.id);
        await this.invokeExperMethod('loadScripts', ...scripts);
    }
    registerFencedCodeRenderer(type, opts) {
        return this.ensureHostScope().logseq.api.exper_register_fenced_code_renderer(this.ctx.baseInfo.id, type, opts);
    }
    registerExtensionsEnhancer(type, enhancer) {
        const host = this.ensureHostScope();
        switch (type) {
            case 'katex':
                if (host.katex) {
                    enhancer(host.katex).catch(console.error);
                }
                break;
            default:
        }
        return host.logseq.api.exper_register_extensions_enhancer(this.ctx.baseInfo.id, type, enhancer);
    }
    ensureHostScope() {
        if (window === top) {
            throw new Error('Can not access host scope!');
        }
        return top;
    }
}
